protomap/proto.go

```diff
@@ -700,6 +726,137 @@ func protoFromPathsInternal(p proto.Message, vals map[*gpb.Path]any, valPrefix,
+		if k.Key == nil {
+			return protoreflect.Value{}, fmt.Errorf("invalid list schema field path does not have a list at the end %s", fieldPath)
+		}
+
+		// Find the parts of the path that are not the list -- we assume that this is 2 elements since
+		// we are in a compressed schema.
+		// TODO(robjs): Currently, this may report incorrectly in an uncompressed schema, but we don't have
+		// a signal to indicate this. One needs to be added to the generated protobufs.
+		keyPath := &gpb.Path{
+			Elem: append(protoPrefix.Elem, absPath.Elem[len(protoPrefix.Elem):len(protoPrefix.Elem)+2]...),
+		}
+		keys = append(keys, k.Key)
+		keyPaths = append(keyPaths, keyPath)
+	}
+
+	le := m.ProtoReflect().NewField(fd).List()
+	seen := []map[string]string{}
+	for i, key := range keys {
+		// Don't try and unmarshal a key that we have already looked at, since there may be multiple paths
+		// that are under the same list member. We map all children of a particular list key when we see it
+		// for the first time.
+		var done bool
+		for _, s := range seen {
+			if reflect.DeepEqual(s, key) {
+				done = true
+				break
+			}
+		}
+		if done {
+			continue
+		}
+		seen = append(seen, key)
+
+		listElemChildren, err := findChildren(vals, valPrefix, keyPaths[i], false, false)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("logic error, error returned from extracting list member children, %v", err)
+		}
+
+		// We now need to create the "XXXKey" message, and populate the key values, subsequent values are then populated
+		// into the one protobuf message field.
+		childMsgEmpty := le.NewElement().Message()
+		childMsgTarget := le.NewElement().Message()
+
+		// Walk through the fields of the XXXKey message that we just created. We use the childMsgEmpty here so that we
+		// don't change the message whilst iterating which causes us to revisit that field. We set the values in
+		// the childMsgTarget message.
+		var retErr error
+		unpopRange{childMsgEmpty}.Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+			// We have one field that is a message in a key message, which is the payload. The remaining fields are the keys.
+			switch fd.Kind() {
+			case protoreflect.MessageKind:
+				m := childMsgTarget.NewField(fd).Message()
+				// We must ignore extra fields from this point in the recursion, because keys map to fields that
+				// are not present in the generated protobuf.
+				if err := protoFromPathsInternal(m.Interface(), listElemChildren, keyPaths[i], keyPaths[i], true); err != nil {
+					retErr = err
+					return false
+				}
+				childMsgTarget.Set(fd, protoreflect.ValueOfMessage(m))
+			default:
+				paths, err := annotatedSchemaPath(fd)
+				if err != nil {
+					retErr = err
+					return false
+				}
+
+				var keyName string
+				for _, p := range paths {
+					keyName = p.Elem[len(p.Elem)-1].Name
+					break
+				}
+				pv, err := listKeyAsProtoValue(fd, key[keyName])
```

Comment:
check and test the case that the key doesn't exist in the path. Alternatively, and I think it's better, to check whether the input paths have all keys supplied in the previous key-gathering loop.

Reply:
To clarify this comment is referring to the case where `keyName` is not in `key`. This can happen if the list of keys in the path is not complete. If this is checked in the previous key-gathering loop then it would be safe to assume that `key[keyName]` always returns a valid value.

Reply:
It's not clear that we can check it beforehand, because we only know what keys there are in the list by the fields in the proto -- but the observation that we were not checking whether the value was actually specified is correct. I added a check for the case that the key is not specified in the list, and the case that we receive a key that we don't map to a field.